src/v/ssx/process.cc

```diff
@@ -0,0 +1,180 @@
+      _cmd_str);
+}
+
+ss::future<std::error_code> process::spawn(
+  const std::filesystem::path& cmd, ss::experimental::spawn_parameters params) {
+    gate_guard g{_gate};
+    if (is_running()) {
+        vlog(proclog.error, "A process is already running: cmd {}", _cmd_str);
+        co_return process_errc::running;
+    }
+
+    // Set command string for logging purposes
+    _cmd_str = cmd.string();
+
+    auto p = co_await ss::experimental::spawn_process(cmd, std::move(params));
+    _process.emplace(std::move(p));
+
+    // Capture output async in the background so the broker is not blocked.
+    ssx::background = ssx::spawn_with_gate_then(_gate, [this] {
+        return consume_input_stream(_process->stdout(), true, _as)
+          .then([this] {
+              return consume_input_stream(_process->stderr(), false, _as);
+          })
+          // According to seastar docs, a process that is not wait'd may leave a
+          // zombie behind. So we do that here.
+          .then([this] { return handle_wait(); })
+          .finally([this] { _process.reset(); });
+    });
+
+    co_return process_errc::success;
+}
+
+ss::future<> process::stop() { co_await _gate.close(); }
+
+process::exit_status_t process::exit_status() {
+    return unwrap_wait_status(_wait_status);
+}
+
+ss::future<> process::handle_wait() {
+    vassert(is_running(), "_process not instantiated");
+
+    _wait_status = co_await _process->wait();
+    auto status = exit_status();
+
+    // There is no signal=0, so an exit_int=0 is the success exit code.
+    if (status.exit_int != 0) {
+        vlog(
+          proclog.error,
+          "Process stop fail: cmd {}, {}={}",
+          _cmd_str,
+          status.exit_reason,
+          status.exit_int);
+    } else {
+        vlog(proclog.trace, "Process stop success: cmd {}", _cmd_str);
+    }
+}
+
+ss::future<std::error_code>
+process::terminate(std::chrono::milliseconds timeout) {
+    gate_guard g{_gate};
+    if (!is_running()) {
+        co_return process_errc::does_not_exist;
+    }
+
+    try {
+        _process->terminate();
+    } catch (const std::system_error&) {
+        // The process is already dead
+        co_return process_errc::does_not_exist;
+    }
+
+    auto deadline = ss::steady_clock_type::now() + timeout;
+
+    co_return co_await ss::with_timeout(
+      deadline,
+      ss::do_until(
+        [this, deadline] {
+            // The callback (do_until loop) is not auto canceled on timeout,
+            // so exit when the deadline is exceeded.
+            if (ss::steady_clock_type::now() > deadline) {
+                return true;
+            }
+
+            return !is_running();
+        },
+        [] { return ss::sleep(std::chrono::milliseconds{5}); }))
+      .then([] {
+          return ss::make_ready_future<process_errc>(process_errc::success);
+      })
+      .handle_exception_type([this](const ss::timed_out_error&) {
+          _process->kill();
```

Comment:
You must check `_process.has_value()` here, otherwise it could race with the background fiber that resets the process and you'll get a bad optional access.